### PR TITLE
dont graph break on list[SymInt] comparisons

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94068
* __->__ #94054

Reland of https://github.com/pytorch/pytorch/pull/92617

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/variables/lists.py [hunk 1]
```diff
@@ -100,6 +102,37 @@ def call_method(
...
+        # but `tuple([2]) != list([2])`
+        if len(left.items) != len(right.items):
+            return ConstantVariable(False, **options)
+        if len(left.items) == 0:
+            return ConstantVariable(True, **options)
+
+        # Generic list comparison works by iterating over left aka self and right the compared-to list.
+        # If we hit here, their lengths are the same and they cannot be expressed as python constants.
+        # So, we iterate over the zipped list items.
+        equal = True
```
Comment (jansel) [inline torch/_dynamo/variables/lists.py:125]: unused?

File: torch/_dynamo/variables/lists.py [hunk 2]
```diff
@@ -100,6 +102,37 @@ def call_method(
 
         return super(BaseListVariable, self).call_method(tx, name, args, kwargs)
 
+    @staticmethod
+    def generic_list_compare(left, tx, op, right, **options):
+        from .builtin import BuiltinVariable
+
+        assert not (
+            left.is_python_constant() and right.is_python_constant()
```
Comment (jansel) [inline torch/_dynamo/variables/lists.py:110]: Will this fail on `()` or `(1,2)`?

Comment (voznesenskym) [reply torch/_dynamo/variables/lists.py:110]: This should get caught in the constant fold - the non dynamic shapes version of the test in this pr (run in test_misc.py vs test_dynamic_shapes.py) goes through

```
 if (
            left.is_python_constant()
            and right.is_python_constant()
            and op in supported_any
        ):
            # constant fold
            self.push(
                ConstantVariable(
                    supported_any[op](
                        left.as_python_constant(), right.as_python_constant()
                    ),
                    **options,
                )
            )
```

File: torch/_dynamo/variables/lists.py [hunk 3]
```diff
@@ -97,6 +99,36 @@ def call_method(
...
+        ), "Illegal generic list compare on constant lists"
+
+        # Most list-like variables implement comparison ops the same way,
+        # so they can re-use this helper.
+        # There are quirks though, like how `tuple([2]) == torch.Size([2])`,
+        # but `tuple([2]) != list([2])`
+        if len(left.items) != len(right.items):
+            return ConstantVariable(False, **options)
```
Comment (jansel) [inline torch/_dynamo/variables/lists.py:115]: @voznesenskym actually I think this is only correct for the `__eq__` op.
```
>>> [True, True] > [True]
True
>>> [] != [True]
True
```

Comment (jansel) [reply torch/_dynamo/variables/lists.py:115]: https://github.com/pytorch/pytorch/pull/94156

**Pull request closed without merging**